src/main/java/net/dries007/tfc/common/container/KnappingContainer.java

```diff
@@ -0,0 +1,199 @@
+    {
+        final ItemStack main = inv.player.getMainHandItem();
+        return main.isEmpty() ? inv.player.getOffhandItem() : main;
+    }
+
+    @Override
+    public void onButtonPress(int buttonID, @Nullable CompoundNBT extraNBT)
+    {
+        matrix.set(buttonID, false);
+
+        if (!hasBeenModified)
+        {
+            if (!player.isCreative() && !consumeAfterComplete)
+            {
+                ItemStack consumedStack = Helpers.consumeItem(this.stack, amountToConsume);
+                if (isOffhand)
+                {
+                    player.setItemInHand(Hand.OFF_HAND, consumedStack);
+                }
+                else
+                {
+                    player.setItemInHand(Hand.MAIN_HAND, consumedStack);
+                }
+            }
+            hasBeenModified = true;
+        }
+
+        // check the pattern
+        Slot slot = slots.get(0);
+        if (slot != null && player.level instanceof ServerWorld)
+        {
+            KnappingRecipe recipe = getMatchingRecipe((ServerWorld) player.level);
+            if (recipe != null)
+            {
+                slot.set(recipe.assemble(this));
+            }
+            else
+            {
+                slot.set(ItemStack.EMPTY);
+            }
+        }
+    }
+
+    @Override
+    protected void addContainerSlots()
+    {
+        addSlot(new RunnableSlot(new ItemStackHandler(1), 0, 128, 46, this::resetMatrix));
+    }
+
+    @Override
+    public void removed(PlayerEntity player)
+    {
+        Slot slot = slots.get(0);
+        ItemStack stack = slot.getItem();
+        if (!stack.isEmpty())
+        {
+            if (!player.level.isClientSide)
+            {
+                ItemHandlerHelper.giveItemToPlayer(player, stack);
+                consumeIngredientStackAfterComplete();
+            }
+        }
+        super.removed(player);
+    }
+
+    /**
+     * Used in client to check a slot state in the matrix
+     * JEI won't cause issues anymore see https://github.com/TerraFirmaCraft/TerraFirmaCraft/issues/718
+     *
+     * @param index the slot index
+     * @return the boolean state for the checked slot
+     */
+    public boolean getSlotState(int index)
+    {
+        return matrix.get(index);
+    }
+
+    /**
+     * Used in client to set a slot state in the matrix
+     * JEI won't cause issues anymore see https://github.com/TerraFirmaCraft/TerraFirmaCraft/issues/718
+     *
+     * @param index the slot index
+     * @param value the value you wish to set the state to
+     */
+    public void setSlotState(int index, boolean value)
+    {
+        matrix.set(index, value);
+    }
+
+    private void resetMatrix()
+    {
+        matrix.setAll(false);
+        requiresReset = true;
+        consumeIngredientStackAfterComplete();
+    }
+
+    @Nullable
+    private KnappingRecipe getMatchingRecipe(ServerWorld level)
+    {
+        return level.getRecipeManager().getRecipeFor(recipeType, this, level).orElse(null);
```

Comment:
Ideally should use the (or a) knapping item cache. Indirect based on a packed version of a int key?

Reply:
later